{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Untitled1.ipynb",
      "provenance": [],
      "authorship_tag": "ABX9TyPfAOuYLAswaU5i1VFM5ivH",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/MusBansal/BNFO601/blob/main/test1_2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import re # python regex module\n",
        "import os.path\n",
        "# Load the genbank file \n",
        "DATA_FILE_GITHUB = \"https://raw.githubusercontent.com/MusBansal/BNFO601/main/BNFO601%202022%20Exam%201/mystery_genome.fasta\"\n",
        "DEFAULT_FILE_NAME = 'mystery_genome.fasta'\n",
        "\n",
        "fileName = DEFAULT_FILE_NAME\n",
        "#Does the file exists locally, if not get it from the github\n",
        "if not os.path.exists(fileName):\n",
        "  #Load the file from Github to the local folder\n",
        "  !wget --no-check-certificate --content-disposition $DATA_FILE_GITHUB\n",
        "   \n",
        "print(\"Reading file:\", fileName)\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ypt4FJ5qIFvf",
        "outputId": "342f3561-60d2-400b-9eb9-fcd6c06bae61"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-02-22 17:06:15--  https://raw.githubusercontent.com/MusBansal/BNFO601/main/BNFO601%202022%20Exam%201/mystery_genome.fasta\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1856309 (1.8M) [text/plain]\n",
            "Saving to: ‘mystery_genome.fasta’\n",
            "\n",
            "mystery_genome.fast 100%[===================>]   1.77M  --.-KB/s    in 0.06s   \n",
            "\n",
            "2022-02-22 17:06:16 (30.4 MB/s) - ‘mystery_genome.fasta’ saved [1856309/1856309]\n",
            "\n",
            "Reading file: mystery_genome.fasta\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "a=\"abcd\"\n",
        "print(len(a))\n",
        "print(a[-1])\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5Fs0qEGXhzEy",
        "outputId": "2da23d11-59cb-4e82-90e0-61406c9b2676"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "4\n",
            "d\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "cut_site_list=[]\n",
        "nested_list=[]\n",
        "frag_list=[]\n",
        "sequence=\"BNNNGAATTCNNNNNNNNNNNNNNNNNNNNGAATTCCNNNNNNNNNNNNGAATTCCNNNNNNNNNNNNGAATTCCNNNNNE\"\n",
        "recognition_sequence=r'GAATTC'\n",
        "bond_position= 1\n",
        "for match in re.finditer(recognition_sequence, sequence):\n",
        "  sGroup = match.group()\n",
        "  sStart = match.start()\n",
        "  sEnd = match.end() + 1\n",
        "  cut_sites = sStart+bond_position \n",
        "  cut_site_list.append(cut_sites)\n",
        "  #print(sGroup, sStart, sEnd, cut_site_list)\n",
        "\n",
        "current_cut_site=cut_site_list[0]\n",
        "#print(cut_site_list[0])\n",
        "#frag_list.append(sequence[0:current_cut_site])\n",
        "#print(frag_list)\n",
        "#print(len(cut_site_list))\n",
        "count=0\n",
        "\n",
        "for site in range(len(cut_site_list)):\n",
        "  count = count + 1 \n",
        "  frag_list.append(count)\n",
        "  if nested_list == []:\n",
        "    frag_list.append(len(sequence[0:cut_site_list[site]]))\n",
        "    frag_list.append(1)\n",
        "    frag_list.append(cut_site_list[site]+1)\n",
        "    frag_list.append(sequence[0:cut_site_list[site]])\n",
        "    nested_list.append(frag_list)\n",
        "    frag_list=[]\n",
        "    pass\n",
        "  else:\n",
        "    frag_list.append(len(sequence[cut_site_list[site-1]:cut_site_list[site]]))\n",
        "    frag_list.append(cut_site_list[site-1] + 1 )\n",
        "    frag_list.append(cut_site_list[site] + 1)\n",
        "    frag_list.append(sequence[cut_site_list[site-1]:cut_site_list[site]])\n",
        "    frag_list=[]\n",
        "  nested_list.append(frag_list)\n",
        "frag_list.append(len(cut_site_list)+1)\n",
        "frag_list.append(len(sequence[cut_site_list[-1]:]))\n",
        "frag_list.append(cut_site_list[-1]+1)\n",
        "frag_list.append(len(sequence))\n",
        "frag_list.append(sequence[cut_site_list[-1]:])\n",
        "#nested_list.append(frag_list)\n",
        "print(len(nested_list))\n",
        "\n",
        "count=0\n",
        "GC=0\n",
        "for nuc in sequence:\n",
        "  count += 1\n",
        "  #print(nuc)\n",
        "  if nuc == \"G\" or nuc == \"C\": \n",
        "    GC += 1\n",
        "  else:\n",
        "    pass\n",
        "GC_count=GC/count\n",
        "#print(GC_count)\n",
        "\n",
        "pattern=\"AATCGG\"\n",
        "prob_of_pattern=1\n",
        "for nuc in pattern:\n",
        "  if nuc == \"G\" or nuc ==\"C\":\n",
        "    prob_of_pattern = prob_of_pattern*GC_count\n",
        "  elif nuc == \"A\" or nuc == \"T\":\n",
        "    prob_of_pattern = prob_of_pattern *(1-GC_count)  \n",
        "#print(prob_of_pattern)\n",
        "expected_number_of_fragments=prob_of_pattern*(len(sequence)-len(pattern)+1)\n",
        "print(expected_number_of_fragments)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YVcmq_0YRk32",
        "outputId": "1f20d8ea-517a-44e8-c080-614a2752d3b0"
      },
      "execution_count": 66,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "5\n",
            "0.12285013965716773\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FIVck-PNHh_I",
        "outputId": "40d18eb9-e948-40db-8232-4914fc2dd06d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Total length of sequence was 1830138 base pairs\n",
            "Cuts occurred at nucleotides \n"
          ]
        }
      ],
      "source": [
        "__author__ = 'Muskan_Bansal'\n",
        "\n",
        "import re\n",
        "\n",
        "class Fragment(object):\n",
        "\n",
        "    \"\"\" I am providing for you here a user-defined type that will keep track of all of the information for a restriction\n",
        "    fragment.  The instance variables and the methods tell (or should) the whole story.  Study the code, there's at\n",
        "    least one nice trick here worth knowing.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, fragment_name_or_number, fragment_length, begin_position, end_position, fragment_sequence):\n",
        "\n",
        "        self.fragment_name_or_number = fragment_name_or_number\n",
        "        self.fragment_length = fragment_length\n",
        "        self.begin_position = begin_position\n",
        "        self.end_position = end_position\n",
        "        self.fragment_sequence = fragment_sequence\n",
        "\n",
        "    def __str__(self): # Special method attribute so that Fragment objects know how to print themselves\n",
        "\n",
        "        return ('>Fragment ' + str(self.fragment_name_or_number) + ' of ' + str(self.fragment_length) + ' NT '\n",
        "                + '(' + str(self.begin_position) + '..' + str(self.end_position) + ')\\n' + self.fragment_sequence)\n",
        "\n",
        "    def __lt__(self, other):        # This method implements a \"less than\" comparison between two Fragment objects\n",
        "                                    # Amazingly, this is the only comparison operator you need to implement to make\n",
        "                                    # a user-defined type sortable using the built-in sort functionality.\n",
        "\n",
        "        return self.fragment_length < other.fragment_length  # Fragment length is being used as the basis for sorting\n",
        "\n",
        "class DNAdigest(object):\n",
        "\n",
        "    \"\"\"\n",
        "    PURPOSE: Open a file with a single DNA sequence in FastA format\n",
        "             Analyze the DNA sequence for restriction endonuclease cut sites\n",
        "             and do various other things specified in the exam materials.\n",
        "    INPUT FILES:\n",
        "             sequence_file: input file containing a single sequence in FA format\n",
        "\n",
        "    OUTPUT FILE:\n",
        "                output file of DNA fragments produced by the digestions\n",
        "                Annotation is fragment number, length, fragment begin, and fragment end positions\n",
        "    \"\"\"\n",
        "\n",
        "\n",
        "    def __init__(self, sequence_file=None, recognition_sequence=r'GAATTC', bond_position=1):\n",
        "\n",
        "        \"\"\"\n",
        "        PURPOSE: The only jobs of the initializer will be to:\n",
        "            1.  Compile a regular expression corresponding to the restriction enzyme recognition sequence, and save this\n",
        "                as an instance variable self.RE_RE (for Restriction Enzyme Regular Expression... cute, right?)\n",
        "            2.  Create an instance variable self.bond_position that keeps track of the position within the recognition\n",
        "                sequence where the phosphodiester bond is actually cut.\n",
        "            3.  Create an instance variable self.cut_positions that for now is an empty list.  Later we will populate it\n",
        "                with positions in the sequence we have read in where cut sites were found.\n",
        "            4. Open the input sequence file, ignore the annotation line, and populate an instance variable self.sequence\n",
        "               for use elsewhere in the program\n",
        "        \"\"\"\n",
        "\n",
        "        self.RE_RE = re.compile(recognition_sequence)\n",
        "        self.bond_position = bond_position\n",
        "        self.sequence = ''\n",
        "        self.cut_positions = []\n",
        "        self.fragment_info = []\n",
        "\n",
        "        if sequence_file:\n",
        "          self.sequence = \"\"        \n",
        "          with open(fileName, 'r') as f:\n",
        "            for line in f:\n",
        "              line = line.strip()\n",
        "              if line.startswith('>'):\n",
        "                pass\n",
        "              else:\n",
        "                self.sequence += line\n",
        "                  \n",
        "        else:\n",
        "          self.sequence = 'BNNNGAATTCNNNNNNNNNNNNNNNNNNNNGAATTCCNNNNNNNNNNNNNNNNNNNNNNNNNNNNNE'\n",
        "            # a sequence containing a couple of EcoRI cut sites, just for testing.\n",
        "\n",
        "        print ('Total length of sequence was', len(self.sequence), 'base pairs')\n",
        "\n",
        "        #print (self.sequence)\n",
        "\n",
        "    def search_for_cuts(self):\n",
        "\n",
        "        \"\"\"\n",
        "        This method will populate the instance variable list self.cut_positions with the locations of the cut sites\n",
        "\n",
        "        Clue 1:  Consider using the finditer method to search your sequence.\n",
        "        Clue 2:  There are useful methods associated with match objects, one of which is ideal for reporting cut sites\n",
        "        Clue 3:  Don't forget to account for the bond position where the cut actually occurs\n",
        "        Clue 4:  It's OK to track cut positions internally \"pythonically\" (i.e. zero indexing) instead of biologically\n",
        "                 (i.e. first residue is numbered 1).  Don't forget to add one though later when you finally report!\n",
        "        \"\"\"\n",
        "      for match in re.finditer(self.RE_RE, self.sequence,self.bond_position):\n",
        "        sStart = match.start()\n",
        "        #sEnd = match.end() + 1\n",
        "        cut_sites = sStart+bond_position \n",
        "        self.cut_positions.append(cut_sites)\n",
        "\n",
        "        #  some stuff here. See clues above.\n",
        "\n",
        "        return self.cut_positions\n",
        "\n",
        "    def report_cuts(self, output_file=None, sort_by_length=True):\n",
        "\n",
        "        \"\"\"\n",
        "        This method should do three things.\n",
        "\n",
        "        First, it should figure out the actual DNA sequence of the fragments\n",
        "        produced.\n",
        "        We are going to simplify here in two ways... first, we assume that the DNA we are cutting is linear,\n",
        "        so we don't have to worry about a fragment that spans position 0.  The second simplifying assumption is to\n",
        "        ignore the sticky ends issue.  So the cut site position should correspond just to the position in the DNA where\n",
        "        the phosphodiester bond was cut (so for a G/AATTC cutter, we would just have a first fragment 5'...NNNNG - 3',\n",
        "        and a second fragment 5' - AATTCNNN... - 3'  etc.).\n",
        "\n",
        "        Second, it should populate a list self.fragment_info with Fragment objects that summarize each fragment.\n",
        "         Fragments should known their fragment number (start from 1), fragment length,\n",
        "         start position in original sequence (this should be the BIOLOGICAL position), the end position in the\n",
        "         original sequence (again, biological sequence coordinates), and finally a string with the actual DNA sequence.\n",
        "\n",
        "        Third, it should either print (if argument output_file=None) or send to a file (if argument output_file is any\n",
        "        other string) in fastA format the fragments.  If argument sort_by_length=True, these should be sorted by\n",
        "        fragment length, otherwise, just the default ordering by fragment number.\n",
        "\n",
        "        \"\"\"\n",
        "\n",
        "        if not self.cut_positions:   # Check and see if the cut position array has been populated.\n",
        "\n",
        "            self.search_for_cuts()   # If not, go ahead and do so by invoking the search_for_cuts method\n",
        "\n",
        "        last_begin = 0\n",
        "        last_end = len(self.sequence)\n",
        "\n",
        "        i = 0\n",
        "        frag_list=[]\n",
        "        for sites in range(len(self.cut_positions)):\n",
        "          i = i + 1 \n",
        "          frag_list.append(i)\n",
        "          if self.fragment_info == []:\n",
        "            frag_list.append(len(self.sequence[0:self.cut_positions[site]]))\n",
        "            frag_list.append(1)\n",
        "            frag_list.append(self.cut_positions[site]+1)\n",
        "            frag_list.append(self.sequence[0:self.cut_positions[site]])\n",
        "            self.fragment_info.append(frag_list)\n",
        "            frag_list=[]\n",
        "            pass\n",
        "          else:\n",
        "            frag_list.append(len(self.sequence[self.cut_positions[site-1]:self.cut_positions[site]]))\n",
        "            frag_list.append(self.cut_positions[site-1] + 1 )\n",
        "            frag_list.append(self.cut_positions[site] + 1)\n",
        "            frag_list.append(self.sequence[self.cut_positions[site-1]:self.cut_positions[site]])\n",
        "            frag_list=[]\n",
        "          self.fragment_info.append(frag_list)\n",
        "        frag_list.append(len(self.cut_positions)+1)\n",
        "        frag_list.append(len(self.sequence[self.cut_positions[-1]:]))\n",
        "        frag_list.append(self.cut_positions[-1]+1)\n",
        "        frag_list.append(len(self.sequence))\n",
        "        frag_list.append(self.sequence[self.cut_positions[-1]:])\n",
        "        #self.fragment_info.append(frag_list)\n",
        "        #print(self.fragment_info)\n",
        "        # some kind of loop to instantiate fragment objects\n",
        "\n",
        "        # Don't forget to append the fragment created by the last cut until the end of the sequence\n",
        "\n",
        "        #  Code to optionally sort the list fragements by fragment length depending on sort_by_length boolean\n",
        "    \n",
        "        if sort_by_length:\n",
        "          self.fragment_info=sorted(nested_list, key=lambda frag:frag[1])\n",
        "\n",
        "        else:\n",
        "          self.fragment_info= sorted(nested_list, key=lambda frag:frag[0])\n",
        "\n",
        "        if output_file:\n",
        "          with open(output_file, \"w\") as of:\n",
        "            for frag in range(self.fragment_info):\n",
        "              print(\">frag#{}|{}|{}\\n{}\".format(self.fragment_info[frag][0],self.fragment_info[frag][2], self.fragment_info[frag][3], self.fragment_info[frag][4]))\n",
        "            pass # something actally meaningful here to write things to an output file\n",
        "\n",
        "        else:\n",
        "\n",
        "            for frag in self.fragment_info:\n",
        "\n",
        "                print (frag)\n",
        "\n",
        "\n",
        "    def observed_and_expected(self):\n",
        "\n",
        "        \"\"\" You'll need to provide some code here mostly on your own. On the \"observed\" side, using this method\n",
        "        presupposes that the genome has already been searched for fragments.\n",
        "        The \"expected\" side of things is a bit more open ended with respect to possible approaches.  At a minumum,\n",
        "        you'll probably also need to do some other type of analysis of the input file, and perhaps of the recognition\n",
        "        sequence you searched for.\n",
        "        \"\"\"\n",
        "        gc_count=self.__calculate_GC(self.sequence)\n",
        "        number_of_observed_fragments=len(self.frag_info)\n",
        "        counting_frag_length= 0\n",
        "        for frag in range(self.fragment_info):\n",
        "          counting_frag_length += self.fragment_info[frag][1]\n",
        "        average_length_of_observed_fragments= counting_frag_length/number_of_observed_fragments\n",
        "\n",
        "        prob_of_pattern=1\n",
        "        for nuc in self.RE_RE:\n",
        "          if nuc == \"G\" or nuc ==\"C\":\n",
        "            prob_of_pattern = prob_of_pattern*gc_count\n",
        "          elif nuc == \"A\" or nuc == \"T\":\n",
        "            prob_of_pattern = prob_of_pattern *(1-gc_count)  \n",
        "        #print(prob_of_pattern)\n",
        "        expected_number_of_fragments=prob_of_pattern*(len(sequence)-len(pattern)+1)\n",
        "        #print(expected_number_of_fragments)\n",
        "        average_length_of_expected_fragements=len(self.sequence)/expected_number_of_fragments\n",
        "        print ('We observed', number_of_observed_fragments,'of average length', average_length_of_observed_fragments)\n",
        "        print (\"We expected\", expected_number_of_fragments, 'of average length', average_length_of_expected_fragements)\n",
        "\n",
        "        print ('Observed over Expected fragment number', number_of_observed_fragments/expected_number_of_fragments)\n",
        "        print ('Observed over Expected average fragment length', average_length_of_observed_fragments/\n",
        "                                                                 average_length_of_expected_fragements)\n",
        "\n",
        "\n",
        "    def __calculate_GC(self, sequence):\n",
        "\n",
        "        \"\"\"\n",
        "        It may turn out to be handy to have a method that calculates the GC percentage.\n",
        "        This private method should take a sequence string as an argument and returns a float corresponding to\n",
        "        GC percentage.\n",
        "        \"\"\"\n",
        "\n",
        "        # do something with sequence\n",
        "        count=0\n",
        "        GC=0\n",
        "        for nuc in self.sequence:\n",
        "          count += 1\n",
        "          #print(nuc)\n",
        "          if nuc == \"G\" or nuc == \"C\": \n",
        "            GC += 1\n",
        "        else:\n",
        "          pass\n",
        "        GC_percent=GC/count\n",
        "        return GC_percent\n",
        "\n",
        "def main():\n",
        "\n",
        "    #myDigest = DNAdigest('mystery_genome.fasta', 'AAGCTT', 1)  ## version for the actual mystery sequence\n",
        "    myDigest = DNAdigest()  # Use the default arguments for quick testing\n",
        "\n",
        "    print ('Cuts occurred at nucleotides', ', '.join(map(str, myDigest.search_for_cuts())))\n",
        "    # if you have properly implemented search for cuts, it returns a list of cut positions.\n",
        "    # The map function here is a handy way to take the numeric values of cut positions and convert them to strings\n",
        "    # for printing here.\n",
        "\n",
        "\n",
        "    myDigest.report_cuts() #  Version for testing that will just print to the screen\n",
        "    #myDigest.report_cuts('YOURNAME_digest_output.txt', False)  # version for producing actual output file\n",
        "\n",
        "\n",
        "if __name__ == '__main__':\n",
        "    main()"
      ]
    }
  ]
}